tests/chapter4_functional_test.py

```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env python3
+
+# Validation code for the chapter-4 examples
+#
+# SPDX-License-Identifier: MIT
+
+import subprocess
+import pytest
+import tempfile
+import os
+import sys
+from pathlib import Path
+from os.path import exists as file_exists
+
+
+def get_script_path():
+    return Path( __file__).parent
+
+def get_root_project_directory():
+    return get_script_path().parent
+
+def get_chapter4_root_directory():
+    return get_root_project_directory().joinpath("chapter_4")
+
+if os.name == 'nt':
+    CMAKE_COMMAND_MULTICONFIG_ARGS=("--config", "Debug")
+    CPACK_COMMAND_MULTICONFIG_ARGS=("-C", "Debug")
+    PLATFORM_EXECUTABLE_EXTENSION=".exe"
+    PLATFORM_STATIC_LIBRARY_PREFIX=""
+    PLATFORM_STATIC_LIBRARY_EXTENSION=".lib"
+else:
+    PLATFORM_EXECUTABLE_EXTENSION=""
+    PLATFORM_STATIC_LIBRARY_PREFIX="lib"
+    PLATFORM_STATIC_LIBRARY_EXTENSION=".a"
+    CMAKE_COMMAND_MULTICONFIG_ARGS=()
+    CPACK_COMMAND_MULTICONFIG_ARGS=()
+
+
+
+CMAKE_COMMAND="cmake"
+CPACK_COMMAND="cpack"
+
+def run_command(command, *args):
+    """Execute a command with arguments.
+
+    Args:
+        command (string): Command to execute
+        *args (args list): Arguments to be forwarded to the
+
+    Returns:
+        bool: True if command execution did not throw any errors, False otherwise
+    """
+    try:
+        subprocess.run(args=[command, *args], check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+        return True
+    except:
+        return False
+
+def run_cmake(*args):
+    """Run CMake with given args
+
+    Returns:
+        bool: True if command execution did not throw any errors, False otherwise
+    """
+    return run_command(CMAKE_COMMAND, *args)
+
+def run_cpack(*args):
+    """Run CPack with given args
+
+    Returns:
+        bool: True if command execution did not throw any errors, False otherwise
+    """
+    return run_command(CPACK_COMMAND, *args)
+
+def test_prerequisites():
+    """
+    Check overall prerequisites for the tests.
+    
+    If one of the checks fail, the testing cannot proceed any further.
+    """
+    assert run_cmake("--version"), "CMake is not present!"
+    assert run_cpack("--version"), "CPack is not present!"
+    # assert run_command("rpmbuild", "--version"), "rpmbuild is not present!"
+    # assert run_command("dpkg", "--version"), "DPKG is not present!"
+    assert run_command("tar", "--version"), "TAR is not present!"
+
+class TestChapter4():
+    """
+    Tests for chapter 4 CMake example content.
+    
+    This code validates that can be configured, built, installed without any issues. In addition,
+    the test code validates installation has installed the expected files (and not installed the unexpected ones) 
+    and tries to run the executables.
+    """
+    temporary_build_root_dir=tempfile.TemporaryDirectory("build")
+    temporary_install_root_dir=tempfile.TemporaryDirectory("install")
```

Comment:
It might be cleaner to create a new install directory for each test in a `setup()` method and delete it again in the `teardown()`. maybe the cmake build can also be in the setup, so that each test case can run independently.

Reply:
Yes, that's the ideal case, but my PyTest skills are not that sharp yet :) I just wrote this with the stuff I learned this morning, so there is plenty of room for improvement :)